### PR TITLE
MGCB Editor dotnet tool

### PR DESCRIPTION
**Pull request by jnoyola**
This creates a dotnet tool "Launcher", `mgcb-editor`, that allows transparent packaging, installation, and launch of the MGCB Editor for each platform.

### Technical Highlights
* `AppReference.targets` added to collect dependencies from .NET Core Apps without taking hard dependencies on them. Comments in the file have more details.
* `FileAssociation.{Platform}.cs` performs registration and unregistration as the old installers did.
* `ResolveDotnetApp` added as a `ProcessStartInfo` extension to resolve the best app available and adjust the parameters.
  * Windows: `{app}.exe {args}`
  * Mac: `open -n {app}.app --args {args}` (or the inner executable if we want to pipe output)
  * Linux: `{app} {args}`
  * Otherwise: `dotnet {app}.dll {args}`

### Validation

* Built on each platform and run on the same platform.

* The package was validated with these steps:

1. On Windows: `dotnet pack -c Release .\Tools\MonoGame.Content.Builder.Editor\MonoGame.Content.Builder.Editor.csproj`
2. On each of Windows, Linux, and Mac
  a. Create nuget.config with local reference
  b. `dotnet new tool-manifest`
  c. `dotnet tool install dotnet-mgcb-editor`
  d. `dotnet mgcb-editor -h` (note: slash options, e.g. `/?`, aren't shown but also work)
  e. `dotnet mgcb-editor`, open project, build
  f. `dotnet mgcb-editor Content.mgcb`, build
  g. `dotnet mgcb-editor --register`, double click project, build (note: this doesn't work on Mac)
  h. `dotnet mgcb-editor --unregister`, double click project to verify it doesn't open with MGCB Editor
  i. `dotnet tool uninstall dotnet-mgcb-editor`

### Known Issues
1. Mac file association is broken. Upon opening a file, it will give an error saying it can't open MGCB files because we don't set the `NSApplication` `OpenFile` delegate. This may require a document-based application. Even if this is fixed, however, Mac apps opened in Finder don't have the same environment variables, so `dotnet` won't be in the PATH for calling `mgcb`. This might need a wrapper like Eto creates for the app itself.
3. Linux icons don't always refresh when calling `gtk-update-icon-cache`

### CODE REVIEW

File: Tools/MonoGame.Content.Builder.Editor/Platform/Linux/FileAssociation.Linux.cs [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        private const string applicationDirectory = "/usr/share/applications";
+        private const string iconRootDirectory = "/usr/share/icons/hicolor";
+        private const string iconType = "scalable/mimetypes";
+
+        // Content files.
+        private const string contentFolder = "Content";
+        private const string iconFileName = "monogame.svg";
+        private const string oldMimetypeFileName = "mgcb.xml";
+        private const string newMimetypeFileName = "x-mgcb.xml";
+        private const string desktopFileName = "MGCB Editor.desktop";
```
Comment (Jjagg) [inline Tools/MonoGame.Content.Builder.Editor/Platform/Linux/FileAssociation.Linux.cs:21]: Probably best not to have a space in here. I think `mgcb-editor.desktop` for the filename would be better.

**Comment by harry-cpp** [comment]
How does the filename association and desktop shortcut work with `dotnet tool install` since it installs the tool in a local location? Shouldn't we enforce the mgcb editor to always be the newest version instead and always be installed globally?

PS. I have a PR incoming that will make mgcb editor use `dotnet mgcb` of the project for content compilation.

**Comment by Jjagg** [comment]
> dotnet mgcb of the project

@harry-cpp What do you mean by this? The version of the MGCB task if there is a reference?

**Comment by harry-cpp** [comment]
MGCB task installs a per project version of dotnet mgcb, not a global version of it. That is what I am trying to make Pipeline Tool use.

**Comment by mrhelmut** [comment]
> MGCB task installs a per project version of dotnet mgcb, not a global version of it. That is what I am trying to make mgcb editor use.

That is cool because it would have the perk of managing mgcb builds for each consoles independently (making mgcb easier to distribute there).

File: Tools/MonoGame.Content.Builder.Editor/Directory.Build.props [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+  <!-- This is needed to differentiate intermediate output directories if you have
+       multiple projects in the same folder and build them in the same solution.
+       See more: https://github.com/NuGet/Home/issues/4463#issuecomment-404806889 -->
+  <PropertyGroup>
+    <MSBuildProjectExtensionsPath>obj\$(MSBuildProjectName)\</MSBuildProjectExtensionsPath>
+  </PropertyGroup>
+
+  <!-- Continue up the folder tree for any additional Directory.Build.props files. -->
+  <Import Project="$([MSBuild]::GetPathOfFileAbove('Directory.Build.props', '$(MSBuildThisFileDirectory)..\'))" />
```
[the remark this answers is not in the document]
Comment (jnoyola) [reply Tools/MonoGame.Content.Builder.Editor/Directory.Build.props:11]: Based on [this](https://github.com/microsoft/msbuild/issues/2456#issuecomment-324532209), they were getting a circular reference back to itself when doing that. It looks like maybe it uses "the directory above" _or_ "the given directory (without going above)"

File: Tools/MonoGame.Content.Builder.Editor/MonoGame.Content.Builder.Editor.Mac.csproj [hunk 3]
```diff
@@ -30,7 +33,7 @@
   </ItemGroup>
 
   <ItemGroup>
-    <None Remove="App.icns" />
+    <None Remove="Icon.icns" />
```
[the remark this answers is not in the document]
Comment (jnoyola) [reply Tools/MonoGame.Content.Builder.Editor/MonoGame.Content.Builder.Editor.Mac.csproj:36]: That PR changed the `BundleResource` a few lines down, but missed this line.

**Comment by Jjagg** [comment]
@harry-cpp so pretty much this: https://github.com/MonoGame/MonoGame/issues/6922#issuecomment-551990096

**Comment by jnoyola** [comment]
> How does the filename association and desktop shortcut work

First, just want to note I didn't add desktop shortcuts, but I can do that if we want it. I can't remember if I forgot it or if I thought we decided we don't want it. 😛

> Shouldn't we enforce the mgcb editor to always be the newest version instead and always be installed globally?

In @Jjagg's link, I believe we agreed whichever one was registered last would stick. I don't think there's any easy way to ensure it's installed globally, and I don't think we'd want to automatically change versions of nuget packages/tools when users expect to be able to choose a specific version.

> MGCB task installs a per project version of dotnet mgcb, not a global version of it. That is what I am trying to make mgcb editor use.

I like that in theory, but I'm not sure how it would work practically. .mgcb files aren't always associated with a single project. What would you do if you open a standalone .mgcb file with the MGCB Editor? I guess you could have a packaged fallback. But what if you have a structure like this?
```
Game
  - Content
    - Content.mgcb
  - Game.DesktopGL
    - Game.DesktopGL.csproj
  - Game.WindowsDX
    - Game.WindowsDX.csproj
```

**Comment by harry-cpp** [comment]
> desktop shortcut work

I meant the Linux .desktop file which is the application launcher XD

> In @Jjagg's link, I believe we agreed whichever one was registered last would stick. I don't think there's any easy way to ensure it's installed globally, and I don't think we'd want to automatically change versions of nuget packages/tools when users expect to be able to choose a specific version.

I was more talking about your installation instructions which require: 

b. dotnet new tool-manifest
c. dotnet tool install dotnet-mgcb-editor

We should instead request the tool to be installed globally which can be done with:

```
dotnet tool install --global dotnet-mgcb-editor
```

And not the weird per project setup.

> I like that in theory, but I'm not sure how it would work practically. .mgcb files aren't always associated with a single project. What would you do if you open a standalone .mgcb file with the MGCB Editor? I guess you could have a packaged fallback. But what if you have a structure like this?

Oh, don't worry about that :)

**Comment by jnoyola** [comment]
Ah ok I think we're in agreement on all those. Yes, I think it will lead to less user confusion if we recommend they install globally.

I'm super interested about what you're thinking for your PR!

File: Tools/MonoGame.Content.Builder.Editor/Platform/Linux/FileAssociation.Linux.cs (hunk 1 above)
Comment (jnoyola) [reply Tools/MonoGame.Content.Builder.Editor/Platform/Linux/FileAssociation.Linux.cs]: Fixed.

**Comment by Jjagg** [comment]
@harry-cpp Is gonna check out the Mac and Linux builds later :)

**Pull request merged**